### ISSUE TITLE
Providing features for specific IDE

### ISSUE THREAD

**Issue by Bukama** [junit-pioneer/junit-pioneer#229]
Today there were two PR at JUnits repository merged that draw @nicolaiparlog s attention and he likes to discuss them:

* [2241 - GitHub issue links in IDEA Git log](https://github.com/junit-team/junit5/pull/2241)
* [2240 - Add icon for JetBrains Toolbox](https://github.com/junit-team/junit5/pull/2240)

As you can read from the link description both things don't target the JUnit code itself, but focus on some improvements (some may call them only gadgets) for IntelliJ.
As they are support only one specific IDE, but not all this would mean, they only are useful for users of this IDE.

So first goal of this issue is do decide if we want to support IDEs with such features in general. If we decide to do so (and how), we then can decide if you want to copy above mentioned PR.

Some general information:

* Inside the `.infra` directory there is already an folder `eclipse` which holds eclipse formatter file and organizing import instruction file. They are natively only usable by eclipse, but for IntelliJ there's a plugin to make use of these files. Root cause of them being in the repository is that they are copied from JUnit Jupiter repository.
* If we are open to support specific IDE it may happen that we will get quite some additional folders inside `.infra` for those. Maybe user expect us to maintain these config files, which would come down to the point if we are able to do this (skill wise)

Let's start :)

**Comment by Bukama** [comment]
The JUnit team was also discussing if they should improve support for specific IDE and they first decline the icon thing, but accepted the link thing, because it saves time for all users of this IDE. After that they reverted their rejection of the icon thing, because their argument "we don't want to add IDE specific configuration files" was gone.

I think (don't know) they changed their mind because those configs may help the contributers of the project. Sure the icon thing is just a visual gimmick (and as @sormuras ) pointed out not even fully supported by IntelliJ itself, the link issue functionality provides real support.

As we already have eclipse specific files in our repository, I'm open to provide config files for other IDE too, especially when they provide support while working. I'm also open for such gadgets but we should be aware not to spend to much time on such. But somehow I even like the icon thing - imagine the pioneer logo would be the JUnit 5 one with ~our~ @nicolaiparlog  pioneer helmet :D

**Comment by sormuras** [comment]
> @nicolaiparlog pioneer helmet

💯

**Comment by nicolaiparlog** [comment]
I really like the idea if improving the project's integration into IntelliJ (since I use it day in day out). I hear @aepfli investigated whether [Gradle's IDEA plugin](https://docs.gradle.org/current/userguide/idea_plugin.html) can do this, which makes sense as this is clearly in its domain - more so than in ours. If we can add to `build.gradle` instead of `.idea/...` I would prefer that. Let's see what conclusion he comes to.

**Comment by nicolaiparlog** [comment]
One thing on the Eclipse config files: They are in there because Spotless uses them ~~to constantly break my PRs~~ to verify the code style. In that sense, they are not so much IDE configuration as build tooling.

**Comment by aepfli** [comment]
As @nicolaiparlog already stated, i would favor to add this to the gradle idea plugin. this plugin is widely used and a default gradle plugin (https://github.com/junit-team/junit5/blob/master/build.gradle.kts#L95 even junit is using it)

So i will try to investigate, if i can configure it, within our project, so it will generate the necessary data. and if not, i mean i have to see, where they host the gradle project - but it would be cool to say i contributed to gradle :). So i will try to take a look at the gradle project, and maybe i can provide a pr there.

**Comment by nicolaiparlog** [comment]
We could also ask @marcphilipp whether the know anything about this (he works for Gradle).

**Comment by Bukama** [comment]
> We could also ask @marcphilipp whether the know anything about this (he works for Gradle).

Or @britter ;)

**Comment by marcphilipp** [comment]
Gradle's core `idea` plugin predates IntelliJ's import feature for Gradle projects. It provides an `idea` task that generates `ipr` files. However, IntelliJ's import reads any [configuration supported by the plugin](https://docs.gradle.org/current/userguide/idea_plugin.html#sec:idea_configuration). There's an additional [idea-ext Gradle plugin](https://github.com/JetBrains/gradle-idea-ext-plugin) by JetBrains which provides a few more options (see wiki pages). However, the `vcs` settings and the icons are part of neither.

**Comment by Bukama** [comment]
> Gradle's core `idea` plugin predates IntelliJ's import feature for Gradle projects. It provides an `idea` task that generates `ipr` files. However, IntelliJ's import reads any [configuration supported by the plugin](https://docs.gradle.org/current/userguide/idea_plugin.html#sec:idea_configuration). There's an additional [idea-ext Gradle plugin](https://github.com/JetBrains/gradle-idea-ext-plugin) by JetBrains which provides a few more options (see wiki pages). However, the `vcs` settings and the icons are part of neither.

As of this I suggest to close this issue as an issue at IDEA / Gradle would be a better way.

**Comment by nicolaiparlog** [comment]
@marcphilipp Thanks for the information. I went back and forth where to create an issue for that:

* Gradle's plugin is the farthest away from such a "nice-to-have" feature
* idea-ext has seen very little development recently
* JetBrains issue tracker makes the least immediate sense (how can IntelliJ read this info if no plugin allows its configuration?)

The latter two just don't seem particularly helpful at the moment, so I think Gradle's plugin would be the best way to get this started. Where are its issues tracked? As part of [Gradle's issues](https://github.com/gradle/gradle/issues)?

**Comment by marcphilipp** [comment]
> Where are its issues tracked? As part of [Gradle's issues](https://github.com/gradle/gradle/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc)?

Yes, since it's a core plugin that's the right place.

**Comment by nicolaiparlog** [comment]
Blocked by gradle/gradle#13004.

**Comment by vlsi** [comment]
Hi there.

I see you are discussing various plugins for Gradle, however, it looks like Gradle overhead is neglected.
One can't execute a Gradle build in order to "figure out the icon for the project". It does not work like that. For instance, if my file manager wants to display an icon, I don't expect it to execute Gradle build for each folder. The same for JetBrains Toolbox: it shows a list of projects, and the projects might even use different build tools. I want the icons to appear fast.

What I think is `project icon` and `issue urls` should be [EditorConfig](https://editorconfig.org/) features. Then it could be better supported across IDEs and tools.

Something like 

```
# .editorconfig
icon_url = .idea/icon.svg
icon_url_dark = .idea/icon_dark.svg
```

**Comment by aepfli** [comment]
https://github.com/editorconfig/editorconfig/issues/425 just for a reference, thank you @vlsi

**Comment by nicolaiparlog** [comment]
@vlsi I've never used the Gradle IDEA plugin so far, so I may be misunderstanding something, but I interpreted the first sentence in the documentation...

> The IDEA plugin generates files that are used by IntelliJ IDEA, thus making it possible to open the project from IDEA (File - Open Project). 

... to mean that I use the plugin to configure IntelliJ-specific options and then run `gradle idea` (once), so it can generate the files in `.idea` for IntelliJ to read.

The issue navigation link and the icon are part of that configuration, so I don't understand why the plugin can't generate those options. I don't expect any external tool to run Gradle - the Toolbox simply reads those files and if I ran `gradle idea` beforehand, everything will be in its place.

**Comment by nicolaiparlog** [comment]
For future reference: The issue navigation link is discussed in JetBrains/gradle-idea-ext-plugin#88.

**Comment by vlsi** [comment]
> to mean that I use the plugin to configure IntelliJ-specific options and then run gradle idea (once), so it can generate the files in .idea for IntelliJ to read.

I don't think that's used often. In practice, IDEA opens build.gradle.* files automatically, and it does much better than Gradle trying to generate project files.

> I don't understand why the plugin can't generate those options

The plugin can generate vcs options, however, that does not scale across IDEs and build systems.
What if I edit the project via both Visual Studio Code and IDEA? Do you suggest to add Gradle configuration for both of them?
What's the point of configuring the icon multiple times?

The same for the build system. What if the project is using Maven?

That is why the short-term solution is to include `.idea/icon.png` as-is, and the long-term solution is to have IDE-neutral and build system-neutral solution like Editorconfig.

**Comment by aepfli** [comment]
For now we will implement a somewhat intermediate solution, where we will do an additional logic after the idea task, which will manipulate the xml files. This will currently only work for intellij idea.

We might want to think about adding this for eclipse too, but for now, we are fine with just intellij

For the logos i am not sure, where we will add the default logos, should we put them into the project root?

**Comment by vlsi** [comment]
> where we will do an additional logic after the idea task

Is `idea` task any better than IDEA's own import?

**Comment by marcphilipp** [comment]
> Is `idea` task any better than IDEA's own import?

No, I think it's rather the other way around.

**Comment by vlsi** [comment]
>No, I think it's rather the other way around.

That is why the following looks suspicious to me:

>where we will do an additional logic **after the idea task**

**Comment by aepfli** [comment]
Maybe i was not good at explaining or writing things down as it was in a hurry. but a prototype would be something like

```
import groovy.xml.XmlUtil

build {
    doLast {
        def ideaFolder = new File('.idea');
        if(ideaFolder.exists() && ideaFolder.isDirectory()) {
            def xmlFile = new File(".idea/vcs.xml")
            def xml = new XmlSlurper().parse(xmlFile)
            xml.component.find {it.@name == 'IssueNavigationConfiguration'}.replaceNode {}
            xml.appendNode {
                component(name: "IssueNavigationConfiguration") {
                    option(name: 'links') {
                        list {
                            IssueNavigationLink {
                                option(name: "issueRegexp", value: '#(\\d+)')
                                option(name: "linkRegexp", value: 'https://github.com/junit-pioneer/junit-pioneer/issues/$1')
                            }
                        }
                    }
                }
            }
            xmlFile.withWriter {out->
                XmlUtil.serialize(xml, out)
            }
        }
    }
}
```
in `buildSrc/build.gradle` - this is currently just a prototype, but working, my plan is to investigate also into a kotlin version, to have just one type of dsl in the project, but this might take a while.

**Comment by seregamorph** [comment]
It seems that solution provided by @aepfli https://github.com/junit-pioneer/junit-pioneer/pull/262 is not working anymore with current versions of IDEA and Gradle.
Neither IDEA reimport nor any Gradle execution is executing this block defined in `buildSrc/build.gradle`
```groovy
build {
        // this line executed
	doLast {
        // this line never executed
```

I'm not quite sure, but in general, Gradle does not support such interceptors because of configuration cache specific (not enabled for this project). E.g. `org.gradle.api.invocation.Gradle` also has `buildFinished` callback which is marked obsolete:
```
     * @deprecated This method is not supported when configuration caching is enabled.
     */
    @Deprecated
    void buildFinished(Action<? super BuildResult> action);
```

Anyway, why don't just add `.idea/vcs.xml` to git like it was done in https://github.com/junit-team/junit5/pull/2241/files ?
Yes, it will create a small issue when trying to pull the branches (you need to delete your local `.idea/vcs.xml` first). It can be especially terrible if you switch to a branch with such file committed and then back - the file will be deleted and IDEA will disable git support (and you need to enable it back or switch the branch in console) :)
But this all will happen once and then just work by default for anyone who will clone this repo.
